# A crash course on Python for Data Analysis

1.	Python basic programming:
    + Python modules or how to extend the instruction set
    + Data Types and Operations.
    + A program in Python
    + A function in Python
    + Slicing
    + References
    + Comprehensions
    + Generators
    + Objects
2.	Advanced programming for data analysis: NumPy.
3.  Advanced programming for data analysis: pandas.
4.  Advanced topics: parallel processing, etc.

<center><img src="files/images/python.jpg"></center>


<center><img src="files/images/monty.jpg"></center>

## Python

The software program that you use to invoke operators is called an **interpreter**. You enter your commands as a ‘dialog’ between you and the interpreter. Commands can be entered as part of a script (a text file with a list of commands to perform) or directly at the *cell*. 

In order to ask to the interpreter what to do, you must **invoke** an operator:

In [13]:
range(10)

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

In [14]:
3 + 4 + 9

16

It’s helpful to think of the computation carried out by an operator as involving four parts:

+ The name of the operator
+ The input arguments
+ The output value
+ Side effects

A typical operation takes one or more input arguments and uses the information in these to produce an output value. Along the way, the computer might take some action: display a graph, store a file, make a sound, etc. These actions are called side effects.

Python is a general-purpose programming language, so when we want to use more specific commands (such as statistical operators or string processing operators) we usually need to import them before we can use them. <br>

> For Scientific Python, one of the most important libraries that we need is **numpy** (Numerical Python), which can be loaded like this: ``import numpy as np``

In [15]:
import numpy as np
np.sqrt(25)

5.0

In [16]:
np.arange(10)

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

Access to the functions, variables and classes of a module depends on the way the module was imported:

In [17]:
import math
math.cos(math.pi)

-1.0

In [18]:
import math as m  # import using an alias
m.cos(m.pi)

-1.0

In [19]:
from math import cos,pi # import only some functions
cos(pi)

-1.0

In [20]:
from math import *   # global import
cos(pi)

-1.0

> It is critical to define a clear policy for ``import`` 

Often the value returned by an operation will be used later on. Values can be stored for later use with the **assignment operator**:

In [21]:
a = 101

The command has stored the value 101 under the name <code>a</code>. Such stored values are called **objects**. 

Making an assignment to an object defines the object. Once an object has been defined, it can be referred to and used in later computations. 

To refer to the value stored in the object, just use the object’s name itself. For instance:

In [22]:
b = np.sqrt(a)
b

10.04987562112089

There are some general rules for object names:

+ Use only letters and numbers and ‘underscores’ (_)
+ Do NOT use spaces anywhere in the name
+ A number cannot be the first character in the name
+ Capital letters are treated as distinct from lower-case letters (i.e., Python is case-sensitive)

In [23]:
3a = 10

SyntaxError: invalid syntax (<ipython-input-23-f986eee6e224>, line 1)

In [24]:
a3 = 10

When you assign a new value to an existing object (*dynamic typing*), the former values of that object is erased from the computer memory. The former value of b was 10.0498756211, but after a new assignment:

In [25]:
b = 'a'
print b

a


The value of an object is changed only via the assignment operator. Using an object in a computation does not change the value. 

The brilliant thing about organizing operators in terms of input arguments and output values is that the output of one operator can be used as an input to another. This lets complicated computations be built out of simpler ones.

One way to connect the computations is by using objects to store the intermediate outputs:

In [26]:
a = np.arange(5)
np.sqrt(a)

array([0.        , 1.        , 1.41421356, 1.73205081, 2.        ])

You can also pass the output of an operator directly as an argument to another operator:

In [27]:
np.sqrt(np.arange(5))

array([0.        , 1.        , 1.41421356, 1.73205081, 2.        ])

### Data Types

Most of the examples used so far have dealt with numbers. But computers work with other kinds of information as well: text, photographs, sounds, sets of data, and so on. The word *type* is used to refer to the kind of information. 

It’s important to know about the types of data because operators expect their input arguments to be of specific types. When you use the wrong type of input, the computer might not be able to process your command.

For our purposes, it’s important to distinguish among several basic types:

+ Numeric (positive and negative) data: 
    + decimal and fractional numbers (**floats**), <code>a = 3.5</code>
    + whole numbers (**integers**), <code> b = -12560</code>, and 
    + arbitrary length whole numbers (**longs**):  <code>c=1809109863596239561236235625629561L</code>
+ **Strings** of textual data - you indicate string data to the computer by enclosing the text in quotation marks (e.g., <code>name = "python"</code>).
+ **Boolean** data: <code>a = True</code> or <code>a = False</code>.
+ **Complex** numbers: <code>a = 2+3j</code>
+ Sequence types: **tuples, lists, sets, dictionaries** and **files**.

In [28]:
a = 10

In [29]:
a = 'a'
print a

a


### Operators

+ Addition (also string, tuple and list concatenation) <code>a + b</code>
+ Subtraction (also set difference): <code>a - b</code>
+ Multiplication (also string, tuple and list replication): <code>a * b</code>
+ Division: <code>a / b</code>
+ Truncated integer division (rounded towards minus infinity): <code>a // b</code>
+ Modulus or remainder: <code>a % b</code>
+ Exponentiation: <code>a ** b</code>
+ Assignment: <code>=</code>, <code>-=</code>, <code>+=</code>,<code>/=</code>,<code>*=</code>, <code>%=</code>, <code>//=</code>, <code>**=</code>
+ Boolean comparisons: <code>==</code>, <code>!=</code>, <code><</code>,<code>></code>,<code><=</code>, <code>>=</code>
+ Boolean operators: <code>and</code>, <code>or</code>, <code>not</code>
+ Membership test operators: <code>in</code>, <code>not in</code>
+ Object identity operators: <code>is</code>, <code>is not</code>
+ Bitwise operators (or, xor, and, complement): <code>|</code>, <code>^</code>, <code>&</code>, <code>~</code>
+ Left and right bit shift: <code><<</code>, <code>>></code>

### Python as a calculator

The Python language has a concise notation for arithmetic that looks very much like the traditional one.

In [30]:
a = 3+2
b= 3.5 * -8
c = 10/6
print a, b, c, 10./6.

5 -28.0 1 1.66666666667


Some math functions are not available in the basic Python module, and they need to be imported from a specific module:

In [31]:
import math   # this instruction is not executed if the module has already been imported
print math.pi + math.sin(100) + math.ceil(2.3)

5.63522701248


### A program in Python

General Rules:

+ All text from a <code>#</code> simbol to the end of a line are considered as comments.
+ Code must be **indented** and sometimes delineated by colons. The Python standard for indentation is four spaces. Never use tabs: it can produce hard to find errors. Set you editor to convert tabs to spaces.
+ Typically, a statement must be on a line. You can use a backslash <code>\</code> at the end of a line to continue a statement on to the next line.


In [32]:
# This program computes the factorial of 100.

fact = 1
n= 100
for factor in range(n,0,-1):
    fact = fact * factor 
print fact    

93326215443944152681699238856266700490715968264381621468592963895217599993229915608941463976156518286253697920827223758251185210916864000000000000000000000000


In [33]:
range(10,0,-1)

[10, 9, 8, 7, 6, 5, 4, 3, 2, 1]

> When we write a colon at the end of an iteration, all lines indented at the next level are considered *part* of the iteration. 

> When we write a line at the same indentation as the iteration, we are closing the iteration.

### A function in Python



#### Factorial

The factorial of a non-negative integer $n$, denoted by $n!$, is the product of all positive integers less than or equal to $n$.  

In [34]:
def factorial(n):
    fact = 1L
    for factor in range(n,0,-1):
        fact = fact * factor
    return fact

In [35]:
factorial(n)

93326215443944152681699238856266700490715968264381621468592963895217599993229915608941463976156518286253697920827223758251185210916864000000000000000000000000L

#### Fibonacci

The Fibonacci Sequence is the series of numbers: 0, 1, 1, 2, 3, 5, 8, 13, 21, 34, ...

The general rule to compute the sequence is very simple: The next number is found by adding up the two numbers before it.

In [36]:
def fib1(n):
    if n==1:
        return 1
    if n==0:
        return 0
    return fib1(n-1) + fib1(n-2)

fib1(20)

# this function cannot compute fib(100)

6765

In [37]:
def fib2(n):
    a, b = 0, 1
    for i in range(1,n+1):
        a, b = b, a + b
    return a

n = 1000
if n<15:
    print fib1(n)
else: 
    print fib2(n)

43466557686937456435688527675040625802564660517371780402481729089536555417949051890403879840079255169295922593080322634775209689623239873322471161642996440906533187938298969649928516003704476137795166849228875


#### Greatest Common Divisor

The greatest common divisor of two positive integers $a$ and $b$ is the largest divisor common to $a$ and $b$.  The Euclidean algorithm, or Euclid's algorithm, is an interative method for computing the greatest common divisor of two integers. 

+ If $a<b$, exchange $a$ and $b$.
+ Divide $a$ by $b$ and get the remainder, $r$. If $r=0$, report $b$ as the GCD of $a$ and $b$.
+ Replace $a$ by $b$ and replace $b$ by $r$. If $r \neq 0$ iterate.



In [38]:
def gcd(a,b): # Euclides algorithm v1.0: pseudocode translation
    r = 1
    while r != 0:
        if a<b:
            c=a
            a=b
            b=c
        r = a%b 
        if r == 0:
            return b
        else:
            a = b
            b = r

gcd(100,16)

4

In [39]:
def gcd(a,b):   # Euclides algorithm v2.0: idiomatic Python
    while a:
        a, b = b%a, a
    return b

gcd(100,16)

4

In [40]:
a = 67
a == False

False

### EXERCISE

Create a function that computes the least common multiple. **Tip**: Re-use the *gcd* function

In [41]:
def lcm(a, b):
    return (a*b)/gcd(a,b)
lcm(a,b)

-1876.0

### String processing with Python

Strings are list of characters:

In [42]:
a = "python"
type(a)

str

In [43]:
print "Hello"

Hello


In [44]:
print "This is 'an example' of the use of quotes and double quotes"
print 'This is "another example" of the use of quotes and double quotes'

This is 'an example' of the use of quotes and double quotes
This is "another example" of the use of quotes and double quotes


We can use the operator ``+`` to concatenate strings:

In [45]:
a = 'He'
b = 'llo'
c = a+b+'!'
print c

Hello!


Substrings within a string can be accessed using **slicing**. Slicing uses ``[]`` to contain the indices of the characters in a string, where the first index is $0$, and the last is $n - 1$ (assuming the string has $n$ characters). 

In [46]:
a = 'Python'
print a[:], a[1], a[2:], a[:3], a[2:4], a[::2], a[1::2]

Python y thon Pyt th Pto yhn


The most advanced string functiona are stored in an external module called ``string``

In [47]:
import string as st
help(st)

Help on module string:

NAME
    string - A collection of string operations (most are no longer used).

FILE
    /home/alourido/anaconda3/envs/py27/lib/python2.7/string.py

MODULE DOCS
    https://docs.python.org/library/string

DESCRIPTION
    Beginning with Python 1.6, many of these functions are implemented as
    methods on the standard string object. They used to be implemented by
    a built-in module called strop, but strop is now obsolete itself.
    
    Public module variables:
    
    whitespace -- a string containing all characters considered whitespace
    lowercase -- a string containing all characters considered lowercase letters
    uppercase -- a string containing all characters considered uppercase letters
    letters -- a string containing all characters considered letters
    digits -- a string containing all characters considered decimal digits
    hexdigits -- a string containing all characters considered hexadecimal digits
    octdigits -- a string containing al

In [48]:
type(st.atof('10.3'))

float

In [49]:
a = 'a'

In [50]:
# press tab
a.

SyntaxError: invalid syntax (<ipython-input-50-745ade49529d>, line 2)

In [51]:
a='Hello'
b = a.lower()
print b

hello


In [52]:
print st.ascii_letters
'a' in st.ascii_letters

abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ


True

### Lists

Lists are a built-in data type which require other data types to be useful. A list is a collection of other objects – floats, integers, complex numbers, strings or even other lists.

Lists also support slicing to retrieve one or more elements. Basic lists are constructed using square braces, ``[]``, and values are separated using
commas.

In [53]:
l=[]
type(l)

list

In [54]:
x=[1,2,3,4,[1,2,3,4],'jordi']
print x[4:], x[0], x[5]

[[1, 2, 3, 4], 'jordi'] 1 jordi


In [55]:
x[-2:]  # The stride can also be negative which can be used to select the
        # elements of a list in reverse order.

[[1, 2, 3, 4], 'jordi']

Lists can be multidimensional and slicing can be done directly in higher dimensions:

In [56]:
x = [[1,2,3,4], [5,6,7,8]]
print x[0], x[0][0]

[1, 2, 3, 4] 1


### Help: Python Tutorial

In [57]:
from IPython.display import HTML
HTML('<iframe src=http://docs.python.org/3/tutorial/index.html?useformat=mobile width=780 height=350></iframe>')

### Conditionals

The conditional structure in Python is <code>if</code>. It is usally combined with 
relational operators: <code> <, <=, ==, >=, >, != </code>.

In [58]:
def main(celsius):
    fahrenheit = 9.0 /5.0 * celsius + 32
    print "The temperature in Fahrenheit is", fahrenheit
    if fahrenheit > 90:
        print "It's really hot out there."
    elif fahrenheit < 30:
        print "It's really cold out there."
    else: pass
        
main(35)

The temperature in Fahrenheit is 95.0
It's really hot out there.


``If`` statesments can be combined with loops (``for``, ``while``):

In [59]:
numbers = [-5, 3,2,-1,9,6]
total = 0
for n in numbers:
    if n >= 0:
        total += n
print total

20


In [60]:
def average(a):
    sum = 0.0
    for i in a:
        sum = sum + i
    return sum/len(a)

average([1,2,3,4])

2.5

In [61]:
def sumdif(x,y):
    sum, dif = x+y, x-y
    return sum, dif

a, b = sumdif(2,2)
print a, b

4 0


In [62]:
def main(n):
    cont = 0
    while (int(n) > 0):
        cont += 1
        n = n/2
#        print n
    return cont-1

main(10)
# main(10.3)

3

### EXERCISE

Analogous function to the previous one, but for product and division

In [69]:
def productdiv(x, y):
    product = x * y
    if y != 0:
        division = x / y
    else:
        division = "Undefined (division by zero)"
    return product, division

a, b = productdiv(6, 2)
print("Product:", a)
print("Division:", b)

c, d = productdiv(8, 0)
print("Product:", c)
print("Division:", d)

('Product:', 12)
('Division:', 3)
('Product:', 0)
('Division:', 'Undefined (division by zero)')


### Boolean operators.

In [70]:
a = 4
b = 40
(a>2) and (b>30)

True

In [71]:
(a>2) or (b>100)

True

In [72]:
not(a>2)

False

### Data Collections

We need to represent data collections: words in a text, students in a course, experimental data, etc., or to store intermediate results. The most simple data collection is the <code>list</code> (an ordered sequence of objects):

In [73]:
range(10)

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

In [74]:
import string
b = string.split("This is an example")
print b

['This', 'is', 'an', 'example']


Lists are *mutable, dynamic and non-homogeneous* objects:

In [75]:
a = [1,2,3,4]
a[1] = 1
print a

[1, 1, 3, 4]


In [76]:
c = a + b
print c

[1, 1, 3, 4, 'This', 'is', 'an', 'example']


In [77]:
zeroes = [0] * 10
del zeroes[5:]
print zeroes

[0, 0, 0, 0, 0]


In [78]:
zeroes.append(1)
print zeroes

[0, 0, 0, 0, 0, 1]


In [79]:
zeroes.remove(1)
print zeroes

[0, 0, 0, 0, 0]


In [80]:
if 1 in zeroes:
    print False
else: 
    print True

True


#### References

We can inspect the reference of an object:

In [81]:
a ='hello'
print id(a)

140604032457872


Two different objects:

In [82]:
a = [1,2,3]
b = [1,2,3]
print id(a), id(b)

140604041710720 140604032174272


Object alias:

In [83]:
a = [1,2,3]
b = a                     # alias
print id(a), id(b)

140604041734416 140604041734416


Cloning:

In [84]:
a = [1,2,3]
b = a[:]                  # cloning with :

print a, b, b[1:], id(a), id(b), id(b[1:])

[1, 2, 3] [1, 2, 3] [2, 3] 140604032282448 140604032174272 140604041735136


When a list is an argument of a function, we are sending the *reference*, not a *copy*

In [85]:
def head(list):
    return list[0]

numbers=[1,2,3,4]
print head(numbers), numbers

1 [1, 2, 3, 4]


In [86]:
def change_first_element(list):
    list[0]=0
    
numbers=[1,2,3,4]
change_first_element(numbers)
print numbers

[0, 2, 3, 4]


If we return a list we are returning a reference:

In [87]:
def tail(list):
    return list[1:]     # we are creating a new list

numbers=[1,2,3,4]
rest = tail(numbers)
print rest, numbers
print id(rest), id(numbers)

[2, 3, 4] [1, 2, 3, 4]
140604029699888 140604041657472


In [88]:
# Press tab
numbers.

SyntaxError: invalid syntax (<ipython-input-88-cfb12d0780d8>, line 2)

Sometimes it is important to perform a *sanity check* about what is doing a pre-defined function:

In [89]:
numbers=[1,2,3,4]
def test(l):
    return l.reverse()

test(numbers)
print numbers, numbers.reverse(), numbers

[4, 3, 2, 1] None [1, 2, 3, 4]


### Dictionaries

A dictionary is a collection that allows the access of an *element* by using a *key*:

In [90]:
dict = {"d": "D", "b":"B", "c":"C"}
dict["d"]

'D'

Dictionaries are **mutable** and **unordered**:

In [91]:
dict["a"]="A"
dict

{'a': 'A', 'b': 'B', 'c': 'C', 'd': 'D'}

In [92]:
dict.has_key("a")

True

In [93]:
del dict["a"]
print dict

{'c': 'C', 'b': 'B', 'd': 'D'}


In [94]:
dict = {"d": "D", "b":"B", "c":"C"}
dict.items()

[('c', 'C'), ('b', 'B'), ('d', 'D')]

### Tuples

Tuples are **non-mutable** lists:

In [95]:
tup = ('a', 'b', 'c')
print type(tup), tup[1:3]

<type 'tuple'> ('b', 'c')


In [96]:
tup[0]='d'

TypeError: 'tuple' object does not support item assignment

**Example**: How to compute the statistics of words in a document.

In [97]:
# We can read a file in a list of line strings or as a string
# -*- coding: latin-1 -*-

text = open("files/text.txt",'r').readlines()
for l in text:
    print l[:-1]  # The last character is a 'new line'

BORGES....... El último poema

Si pudiera vivir nuevamente mi vida,
en la próxima trataría de cometer más errores.
No intentaría ser tan perfecto, me relajaría más.
Sería más tonto de lo que he sido, de hecho
tomaría muy pocas cosas con seriedad.
Sería menos higiénico.
Correría más riesgos, haría más viajes,
contemplaría más atardeceres, subiría más montañas, nadaría más ríos.
Iría a más lugares adonde nunca he ido,
Comería más helados y menos habas,
tendría más problemas reales
y menos imaginarios

Yo fui una de esas personas que vivió sensata y prolíficamente cada minuto de su vida;
claro que tuve momentos de alegría.
Pero si pudiera volver atrás trataría de tener
solamente buenos momentos.

Por si no lo saben, de eso está hecha la vida, sólo de momentos;
no te pierdas el ahora.
Yo era uno de esos que nunca iba a ninguna parte sin un termómetro, una bolsa de agua caliente, un paraguas y un paracaídas;
Si pudiera volver a vivir comenzaría a andar descalzo a principios de la primavera


In [98]:
import string

def compare((w1,c1),(w2,c2)):
# A sorting funtion returns negative if x<y, zero if x==y, positive if x>y.
    if c1 < c2:
        return -1
    elif c1 == c2:
        return 0
    else:
        return 1
    
def main():
    
# We read the file in a string
    text = open("files/text.txt",'r').read()
    text = string.lower(text)
    for ch in '!"#$%&/()=?¿^*`+¨[]{}-:;,.':
        text = string.replace(text, ch, ' ')
# We build a list with all words (blank separation)
    words = string.split(text)

# We use a dictionary for counting words
    counts = {}
    for w in words:
        try:
            counts[w] = counts[w] + 1
        except KeyError:
            counts[w] = 1
            
# We create a sorted list
    items = counts.items()
    items.sort(compare)
    
# We print the list
    for i in range(len(items)):
        print "%-2s%3d" % items[i],'|' ,

main()    

termómetro  1 | tener  1 | comería  1 | uno  1 | estoy  1 | sé  1 | relajaría  1 | otra  1 | personas  1 | ven  1 | te  1 | atardeceres  1 | sensata  1 | buenos  1 | ya  1 | agua  1 | helados  1 | perfecto  1 | contemplaría  1 | bolsa  1 | cosas  1 | ninguna  1 | sido  1 | en  1 | ser  1 | cometer  1 | pocas  1 | atrás  1 | está  1 | prolíficamente  1 | último  1 | 85  1 | nuevamente  1 | muy  1 | errores  1 | así  1 | riesgos  1 | eso  1 | higiénico  1 | poema  1 | descalzo  1 | subiría  1 | problemas  1 | luis  1 | pierdas  1 | vivió  1 | años  1 | ahora  1 | otoño  1 | concluir  1 | comenzaría  1 | claro  1 | paracaídas  1 | vez  1 | nadaría  1 | andar  1 | cada  1 | tonto  1 | principios  1 | seguiría  1 | viajes  1 | esos  1 | esas  1 | solamente  1 | lugares  1 | reales  1 | seriedad  1 | saben  1 | intentaría  1 | primavera  1 | iba  1 | mi  1 | tengo  1 | era  1 | habas  1 | tuviera  1 | ríos  1 | fui  1 | tendría  1 | iría  1 | adonde  1 | paraguas  1 | tan  1 | muriendo  1 | 

### Lists (and dictionary) comprehensions

Lists comprehensions are a way to fit a ``for`` loop, an ``if`` statement, and an assignment all in one line.

A list comprehension consists of the following parts:

+ An input sequence.
+ A variable representing members of the input sequence.
+ An optional expression.
+ An output expression producing elements of the output list from members of the input sequence that satisfy the predicate.

In [99]:
num = [1, 4, -5, 10, -7, 2, 3, -1]
squared = [ x**2 for x in num if x > 0]
print type(squared), squared

<type 'list'> [1, 16, 100, 4, 9]


> There is a downside to list comprehensions: the entire list has to be stored in memory at once. This isn’t a problem for small lists like the ones in the above examples, or even of lists several orders of magnitude larger. But we can use **<font color="red">generators</font>** to solve this problem.

Generator expressions do not load the whole list into memory at once, but instead create a *generator object* so only one list element has to be loaded at any time.

Generator expressions have the same syntax as list comprehensions, but with parentheses around the outside instead of brackets:

In [100]:
num = [1, 4, -5, 10, -7, 2, 3, -1]
squared = ( x**2 for x in num if x > 0 )
print type(squared), squared

<type 'generator'> <generator object <genexpr> at 0x7fe0ed3f3f50>


The elements of the generator must be accessed by an iterator because they are generated when needed:

In [101]:
lis = []
for item in squared:
    lis = lis + [item]
print lis

[1, 16, 100, 4, 9]


We can define our own generators with the ``yield`` statesment. For example, let's build a generator for the binary representation of a number between 0 and 1 with arbitrary precision.

In [102]:
# binary representation of a number between 0 and 1 (b bits precision).

def res(n,b):
    bin_a = '.'
    for i in range(b):
        n *= 2
        bin_a +=  str(int(n))
        n = n % 1
    return bin_a

print res(1/3.0,10)

.0101010101


In [103]:
# binary representation of a number between 0 and 1 (precision as needed).

def binRep(n):
    while True:
        n *= 2
        yield int(n)
        n = n % 1

a = binRep(1/3.) 
a_bin = '.'
for i in range(50):
    a_bin +=  str(a.next())
    
print a_bin

.01010101010101010101010101010101010101010101010101


### Objects

You can define your own classes and objects.

In [104]:
#creating a class

class Rectangle:
    def __init__(self,x,y):
        self.x = x
        self.y = y
    description = "This shape has not been described yet"
    author = "Nobody has claimed to make this shape yet"
    def area(self):
        return self.x * self.y
    def perimeter(self):
        return 2 * self.x + 2 * self.y
    def describe(self,text):
        self.description = text
    def authorName(self,text):
        self.author = text
    def scaleSize(self,scale):
        self.x = self.x * scale
        self.y = self.y * scale

#creating objects
a = Rectangle(100, 45)
b = Rectangle(10,230)

#describing the rectangles
a.describe("A fat rectangle")
b.describe("A thin rectangle")

In [105]:
#finding the area of your rectangle:
print a.area()
 
#finding the perimeter of your rectangle:
print a.perimeter()

#getting the description
print a.description
print a.author

4500
290
A fat rectangle
Nobody has claimed to make this shape yet


In [106]:
#finding the area of your rectangle:
print b.area()
print b.description

#making the rectangle 50% smaller
b.scaleSize(0.5)
b.describe("A small thin rectangle")
 
#re-printing the new area of the rectangle
print b.area()
print b.description

2300
A thin rectangle
575.0
A small thin rectangle


## Advanced programming for data analysis: NumPy.


> **NumPy** is an open-source add-on module to Python that provide common mathematical and numerical routines in pre-compiled, fast functions.

There are several ways to import NumPy. The standard approach is: 

In [107]:
import numpy as np

The central feature of NumPy is the **array** object class. Arrays are similar to lists in Python, except that every element of an array must be of the same type, typically a numeric type like float or int. 

Arrays make operations with large amounts of numeric data very fast and are generally much more efficient than lists. 

### Array creation

There are several ways for creating an array:
    
+ Explicitly from a list of values.
+ As a range of values.
+ As a random array.
+ By specifying the number of elements.
+ By creating an unitializated, a zero-initializated or one-initializated array.
+ By creating a constant diagonal value array.
+ Etc.



In [108]:
a = np.array([1, 2, 4, 8, 16], float) 
print a[:2], len(a) 

[1. 2.] 5


In [109]:
print np.arange(5, dtype=float)  
print np.ones((2,3), dtype=float)  
print np.zeros(7, dtype=int)
a= np.array([[1, 2, 3], [4, 5, 6]], float) 
print np.zeros_like(a) 
print np.identity(4, dtype=str) 

[0. 1. 2. 3. 4.]
[[1. 1. 1.]
 [1. 1. 1.]]
[0 0 0 0 0 0 0]
[[0. 0. 0.]
 [0. 0. 0.]]
[['1' '' '' '']
 ['' '1' '' '']
 ['' '' '1' '']
 ['' '' '' '1']]


In [110]:
a = np.linspace(0,100,10) # Return evenly spaced numbers over a specified interval
print a

[  0.          11.11111111  22.22222222  33.33333333  44.44444444
  55.55555556  66.66666667  77.77777778  88.88888889 100.        ]


In [111]:
a = np.linspace(10,100,10)
print a

[ 10.  20.  30.  40.  50.  60.  70.  80.  90. 100.]


In [112]:
np.random.rand(2,3) 

array([[0.21557862, 0.61262031, 0.60238274],
       [0.64997523, 0.36573084, 0.15754381]])

In [113]:
a = np.random.normal(size=5) 
b = np.random.normal(5, size=5)         # mean = 5
print a
print b

[-1.38437439  0.10187906  0.39927713  0.30528785 -2.00432995]
[3.97960585 4.19382889 5.03082807 5.22159763 5.1477999 ]


In [114]:
np.random.normal?

Arrays can be multidimensional. **Unlike lists**, different axes are accessed using commas inside 
bracket notation.

In [115]:
a = np.array([[1, 2, 3], [4, 5, 6]], int)
print a[0,0]                # different axes are accessed using commas inside bracket notation
print a[-1:,-2:]

1
[[5 6]]


Object arrays have several properties:

In [116]:
print a.shape, a.dtype

(2, 3) int64


The ``in`` statement can be used to test if values are present in an array: 


In [117]:
2 in a 

True

Arrays can be reshaped using tuples that specify new dimensions. 

In [118]:
a = a.reshape((3, 2))    # this function creates a new array!
print a

[[1 2]
 [3 4]
 [5 6]]


### EXERCISE

Add a cell and create a new array with a random normal distribution and shape(2,1), and reshape it to (1,2)

In [119]:
a = np.random.normal(size=(2,1))
print a.reshape((1,2)) 

[[-0.21761801  1.306639  ]]


Lists can also be created from arrays: 

In [120]:
a = np.array([1, 2, 3], float) 
type(a.tolist())

list

Other functions and properties:

In [121]:
a = np.array(range(6), float).reshape((2, 3)) 
a.transpose()   # Transposed versions of arrays

array([[0., 3.],
       [1., 4.],
       [2., 5.]])

In [122]:
a.flatten()    # One-dimensional versions of multi-dimensional arrays

array([0., 1., 2., 3., 4., 5.])

In [123]:
a = np.array([1,2], float) 
b = np.array([3,4,5,6], float) 
c = np.array([7,8,9], float) 
np.concatenate((a, b, c))          # Concatenation

array([1., 2., 3., 4., 5., 6., 7., 8., 9.])

In [124]:
a = np.array([[1, 2], [3, 4]], float) 
b = np.array([[5, 6], [7,8]], float) 
np.concatenate((a,b), axis=0)      # it is possible to specify the axis for concatenation

array([[1., 2.],
       [3., 4.],
       [5., 6.],
       [7., 8.]])

In [125]:
np.concatenate((a,b), axis=1)

array([[1., 2., 5., 6.],
       [3., 4., 7., 8.]])

In [126]:
a = np.array([2, 4, 3], float) 
print a.sum(), a.prod(), a.mean(), a.var(), a.std(), a.min(), a.max()
print a.argmin(), a.argmax()     # these functions return the array indices 

9.0 24.0 3.0 0.6666666666666666 0.816496580927726 2.0 4.0
0 1


One dimensional arrays have a 1-tuple for their shape: <code> (size,)</code>.

N-dimensional arrays have a N-tuple for their shape: <code> (size1, ..., sizeN)</code>.

For multidimensional arrays, each of the functions thus far described can take an optional 
argument ``axis`` that will perform an operation along only the specified axis, placing the results in a return array. 

In [127]:
a = np.array([[0, 2], [3, -1], [3, 5]], float) 
print a
print a.mean(axis=0), a.mean(axis=1), a.min(axis=1), a.max(axis=0) 

[[ 0.  2.]
 [ 3. -1.]
 [ 3.  5.]]
[2. 2.] [1. 1. 4.] [ 0. -1.  3.] [3. 5.]


<div class="alert alert-info">`ipythonblocks` is a teaching tool that allows students to experiment with Python flow control concepts and immediately see the effects of their code represented in a colorful, attractive way. BlockGrid objects can be **indexed and sliced like 2D NumPy arrays** making them good practice for learning how to access arrays. </div>


In [128]:
import os
os.chdir('./modules/')
from ipythonblocks import BlockGrid
os.chdir('..')
os.getcwd()

'/home/alourido/Escritorio/Master_de_big_data_USC/Unstructured_information_management_technologies/practices/1_python_introduction'

In [129]:
grid = BlockGrid(8, 8, fill=(123, 234, 123))
grid

In [130]:
grid[0, 0] = (0, 0, 0)
grid[0, 2] = (255, 0, 0)
grid[0, 4] = (255, 255, 255)
grid[0, 6] = (0, 150, 150)
grid.show()

In [131]:
from ipythonblocks import colors
grid[1, 1] = colors['Teal']
grid[1, 2] = colors['Thistle']
grid[1, 3] = colors['Peru']
grid.show()

In [132]:
a = grid[0, 0]
a

Block(0, 0, 0, size=20)

In [133]:
print a

Block [0, 0]
Color: (0, 0, 0)


In [134]:
print grid[0,0] == grid[0,1]    
print grid[0,1] == grid[0,3]

False
True


### Indexing and Slicing

Array views contain a pointer to the original data, but may have different
shape or stride values.


> **Simple assigments do not make copies of arrays. Slicing operations do not make copies either; they return views on the original array**.

In [135]:
a = np.array(range(64)).reshape((8,8))
print a

[[ 0  1  2  3  4  5  6  7]
 [ 8  9 10 11 12 13 14 15]
 [16 17 18 19 20 21 22 23]
 [24 25 26 27 28 29 30 31]
 [32 33 34 35 36 37 38 39]
 [40 41 42 43 44 45 46 47]
 [48 49 50 51 52 53 54 55]
 [56 57 58 59 60 61 62 63]]


In [136]:
print a[0:2,:] 

[[ 0  1  2  3  4  5  6  7]
 [ 8  9 10 11 12 13 14 15]]


In [137]:
grid = BlockGrid(8, 8, fill=(123, 234, 123))
grid[0:2,:] = colors['Teal']
grid.show()

In [138]:
print a[2,1:]  # This is an element!

[17 18 19 20 21 22 23]


In [139]:
grid[2,1:] = colors['Teal']
grid.show()

In [140]:
def eras(n):
  siv=range(n+1)
  # 1 is not prime
  siv[1]=0
  # The last possible factor is sqrt(n)
  sqn=int(round(n**0.5))
  for i in range(2,sqn+1):
    if siv[i]!=0:
        siv[2*i:n/i*i+1:i]=[0]*(n/i-1)
  return filter(None,siv)

index = eras(5000)

col=50
row=100
grid = BlockGrid(col, row, block_size=10, fill=(123, 234, 123))
grid[0,0]=(0,0,0)
grid[0,1]=(0,0,0) 

for i in index:
    grid[i/col,i%col]=[0,0,0]

grid

### Array operations

Standard mathematical operations are applied on an **element-by-element basis**.

In [141]:
a = np.array([1,2,3], float) 
b = np.array([5,2,6], float) 
print a + b, a % b, a ** b

[6. 4. 9.] [1. 0. 3.] [  1.   4. 729.]


For two-dimensional arrays, multiplication remains elementwise and does not correspond to 
matrix multiplication. There are special functions for matrix math. 

In [142]:
a = np.array([[1,2], [3,4]], float) 
b = np.array([[2,0], [1,3]], float) 
a * b 

array([[ 2.,  0.],
       [ 3., 12.]])

Arrays that do not match in the number of dimensions will be **broadcasted** by Python 
to perform mathematical operations. This often means that the smaller array will be repeated 
as necessary to perform the operation indicated.

In [143]:
a = np.array([[1, 2], [3, 4], [5, 6]], float) 
b = np.array([-1, 3], float) 
print a , b
print 
print a + b 

[[1. 2.]
 [3. 4.]
 [5. 6.]] [-1.  3.]

[[0. 5.]
 [2. 7.]
 [4. 9.]]


Sometimes, however, we can use the ``newaxis`` constant to specify how we 
want to broadcast:

In [144]:
a = np.zeros((2,2), float) 
b = np.array([-1., 3.], float) 
print a, b
print
print a + b 
print
print a + b[np.newaxis,:] 
print
print a + b[:,np.newaxis] 


[[0. 0.]
 [0. 0.]] [-1.  3.]

[[-1.  3.]
 [-1.  3.]]

[[-1.  3.]
 [-1.  3.]]

[[-1. -1.]
 [ 3.  3.]]


NumPy offers a large library of common mathematical functions that can be applied elementwise to arrays. Among these are the functions: <code> abs,sign, sqrt, log, log10, exp, sin, cos, tan, arcsin, arccos, arctan, sinh, cosh, tanh, arcsinh, arccosh, </code> and <code>arctanh </code>. 


In [145]:
a = np.array([1, 4, 9], float) 
np.sin(a)

array([ 0.84147098, -0.7568025 ,  0.41211849])

It is possible to iterate over arrays in a manner similar to that of lists: 

In [146]:
a = np.array([[1, 2], [3, 4], [5, 6]], float) 
for x in a: 
    print x

[1. 2.]
[3. 4.]
[5. 6.]


In [147]:
for [x, y] in a:             # Multiple assignment can also be used with array iteration
    print x * y

2.0
12.0
30.0


In [148]:
a = np.array([6, 2, 5, -1, 0], float) 
a.sort() 
print a

[-1.  0.  2.  5.  6.]


### Comparison operators and value testing 
Boolean comparisons can be used to compare members elementwise on arrays of equal size. 


In [149]:
a = np.array([1, 3, 0], float) 
b = np.array([0, 3, 2], float) 
print a > b 
print a == b 
print a <= b 


[ True False False]
[False  True False]
[False  True  True]


Arrays can be compared to single values using broadcasting: 

In [150]:
a = np.array([1, 3, 0], float) 
a > 2

array([False,  True, False])

The <code>any</code> and <code>all</code> operators can be used to determine whether or not any or all elements of a 
Boolean array are true: 

In [151]:
c = np.array([ True, False, False], bool) 
print any(c), all(c)

True False


Compound Boolean expressions can be applied to arrays on an element-by-element basis using 
special functions ``logical_and``, ``logical_or``, and ``logical_not``. 

In [152]:
a = np.array([1, 3, 0], float) 
print np.logical_and(a > 0, a < 3) 
b = np.array([True, False, True], bool) 
print np.logical_not(b) 
c = np.array([False, True, False], bool) 
print np.logical_or(b, c) 


[ True False False]
[False  True False]
[ True  True  True]


The ``where`` function forms a new array from two arrays of equivalent size using a Boolean filter  to choose between elements of the two. Its basic syntax is: <br>
<code>where(boolarray, truearray, falsearray)</code>

In [153]:
a = np.array([1, 3, 0], float) 
b = np.where(a != 0, 1 / a, 1) 
print b


[1.         0.33333333 1.        ]


/home/alourido/anaconda3/envs/py27/lib/python2.7/site-packages/ipykernel_launcher.py:2: RuntimeWarning: divide by zero encountered in divide
  


### Array item selection and manipulation 
We can use array selectors to **filter** for specific subsets of elements of other arrays. 

In [154]:
a = np.array([[6, 4], [5, 9]], float) 
a >= 6

array([[ True, False],
       [False,  True]])

In [155]:
a[a >= 6] 

array([6., 9.])

In [156]:
a = np.array([[6, 4], [5, 9]], float) 
sel = (a >= 6) 
a[sel] 

array([6., 9.])

It is also possible to select using integer arrays that represent indexes.

In [157]:
a = np.array([2, 4, 6, 8], float) 
b = np.array([0, 0, 1, 3, 2, 1], int)  # the 0th, 0th, 1st, 3rd, 2nd, and 1st elements of a
a[b] 

array([2., 2., 4., 8., 6., 4.])

For multidimensional arrays, we have to send multiple one-dimensional integer arrays to the 
selection bracket, one for each axis.

In [158]:
a = np.array([[1, 4], [9, 16]], float) 
b = np.array([0, 0, 1, 1, 0], int) 
c = np.array([0, 1, 1, 1, 1], int) 
a[b,c] 


array([ 1.,  4., 16., 16.,  4.])

### Vector and matrix mathematics

In [159]:
a = np.array([1, 2, 3], float) 
b = np.array([0, 1, 1], float) 
np.dot(a, b) 


5.0

In [160]:
a = np.array([[0, 1], [2, 3]], float) 
b = np.array([2, 3], float) 
c = np.array([[1, 1], [4, 0]], float) 
print np.dot(b, a)
print np.dot(a, c) 

[ 6. 11.]
[[ 4.  0.]
 [14.  2.]]


It is also possible to generate inner, outer, and cross products of matrices and vectors.

In [161]:
a = np.array([1, 4, 0], float) 
b = np.array([2, 2, 1], float) 
np.outer(a, b) 

print np.inner(a, b) 

print np.cross(a, b) 

10.0
[ 4. -1. -6.]


NumPy also comes with a number of built-in routines for linear algebra calculations. 

In [162]:
a = np.array([[4, 2, 0], [9, 3, 7], [1, 2, 1]], float)
np.linalg.det(a)                 # These can be found in the sub-module linalg

-48.00000000000003

In [163]:
vals, vecs = np.linalg.eig(a)
print vals
print
print vecs

[ 8.85591316  1.9391628  -2.79507597]

[[-0.3663565  -0.54736745  0.25928158]
 [-0.88949768  0.5640176  -0.88091903]
 [-0.27308752  0.61828231  0.39592263]]


In [164]:
b = np.linalg.inv(a) 
print b

[[ 0.22916667  0.04166667 -0.29166667]
 [ 0.04166667 -0.08333333  0.58333333]
 [-0.3125      0.125       0.125     ]]


## Advanced programming for data analysis: pandas.

pandas is a fundamental high-level building block for doing practical, real world data analysis in Python.

pandas is well suited for:

+ Tabular data with heterogeneously-typed columns, as in an SQL table or Excel spreadsheet
+ Ordered and unordered (not necessarily fixed-frequency) time series data.
+ Arbitrary matrix data (homogeneously typed or heterogeneous) with row and column labels.


> **pandas is a Python package providing fast, flexible, and expressive data structures designed to work with both relational or labeled data. **.


Key features:

+ Easy handling of missing data
+ Size mutability: columns can be inserted and deleted from DataFrame
+ Powerful, flexible ``group by`` functionality to perform split-apply-combine operations on data sets
+ Intelligent label-based slicing, fancy indexing, and subsetting of large data sets
+ Intuitive merging and joining data sets
+ Flexible reshaping and pivoting of data sets
+ Hierarchical labeling of axes
+ Robust IO tools for loading data from flat files, Excel files, databases, and HDF5
+ Time series functionality: date range generation and frequency conversion, moving window statistics, moving window linear regressions, date shifting and lagging, etc.

Pandas has a lot of functionality, so we'll only be able to cover a **small** fraction of what you can do. Check out the (very readable) pandas docs if you want to learn more:

http://pandas.pydata.org/pandas-docs/stable/

### Series and DataFrames

A Series is a single vector of data with an index that labels every element in the vector. 

If we do not specify the index, a sequence of integers is assigned as index.

In [165]:
import pandas as pd                            # convention, alias 'pd'
c = pd.Series([1956, 1967, 1989, 2000])
c

0    1956
1    1967
2    1989
3    2000
dtype: int64

Its values are stores in a NumPy array (``values``) and the index in a pandas ``Index`` object:

In [166]:
c.values

array([1956, 1967, 1989, 2000])

In [167]:
c.index

RangeIndex(start=0, stop=4, step=1)

In [168]:
c = pd.Series([1956, 1967, 1989, 2000], index = ['a','b','c','d'])
c

a    1956
b    1967
c    1989
d    2000
dtype: int64

In [169]:
print c['d'], c[3]

2000 2000


DataFrames are designed to store heterogeneous multivariate data.

In [170]:
c = pd.DataFrame({'VarA':['aa','bb'], 'VarB':[22.2,33.3]}, \
                 index = ['Case1','Case2'])
c

VarA  VarB
Case1   aa  22.2
Case2   bb  33.3

### EXERCISE:

Create your own dataframe from this dict: {'Player':['Messi','Ronaldo'], 'NGoals':[43,39]}

In [171]:
data = {'Player': ['Messi', 'Ronaldo'], 'NGoals': [43, 39]}
pd.DataFrame(data)

NGoals   Player
0      43    Messi
1      39  Ronaldo

A DataFrame has a second index representing columns:

In [172]:
c.columns

Index([u'VarA', u'VarB'], dtype='object')

We can access **columns** as in a dictionary or with the ``.`` notation:

In [173]:
c.VarA

Case1    aa
Case2    bb
Name: VarA, dtype: object

In [174]:
c['VarA']

Case1    aa
Case2    bb
Name: VarA, dtype: object

These two methods return a ``Series`` object:

In [175]:
type(c.VarA), type(c['VarA'])

(pandas.core.series.Series, pandas.core.series.Series)

If we want to access a **row** in a DataFrame we can index its ``ix`` attribute:

In [176]:
print c.ix[0]
print c.ix['Case1']

VarA      aa
VarB    22.2
Name: Case1, dtype: object
VarA      aa
VarB    22.2
Name: Case1, dtype: object


/home/alourido/anaconda3/envs/py27/lib/python2.7/site-packages/ipykernel_launcher.py:1: DeprecationWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#ix-indexer-is-deprecated
  """Entry point for launching an IPython kernel.
/home/alourido/anaconda3/envs/py27/lib/python2.7/site-packages/ipykernel_launcher.py:2: DeprecationWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#ix-indexer-is-deprecated
  


Or to use the ``irow`` *function*, that let's you grab the ith row from a DataFrame:

In [177]:
print c.irow(0)

AttributeError: 'DataFrame' object has no attribute 'irow'

In [178]:
print c.iloc[0]

VarA      aa
VarB    22.2
Name: Case1, dtype: object


### Reading tabular data from a file

The ‘pandas’ Python library provides several operators, <code>read_csv(), read_tab(), ...</code> that allows you to access data files in tabular format on your computer as well as data stored in web repositories.

Reading in a data table is simply a matter of knowing the name (and location) of the data set.

In [179]:
import pandas as pd
import numpy as np

# Set some Pandas options
pd.set_option('html', False)
pd.set_option('max_columns', 30)
pd.set_option('max_rows', 20)

OptionError: 'Pattern matched multiple keys'

In [180]:
import pandas as pd
import numpy as np

# Set some Pandas options
pd.set_option('display.html.table_schema', False)
pd.set_option('max_columns', 30)
pd.set_option('max_rows', 20)

In [181]:
import pandas as pd
data = pd.read_csv("http://www.mosaic-web.org/go/datasets/swim100m.csv")
data.shape   # an attribute to see how many cases and variables there are in a data frame

(62, 3)

In [182]:
type(data)

pandas.core.frame.DataFrame

In [183]:
data.head()  # display of the firsts rows

year  time sex
0  1905  65.8   M
1  1908  65.6   M
2  1910  62.8   M
3  1912  61.6   M
4  1918  61.4   M

In [184]:
# Load car dataset
auto = pd.read_csv("data/Auto.csv")
auto.tail()  # print the last lines

mpg  cylinders  displacement horsepower  weight  acceleration  year  \
392  27.0          4         140.0         86    2790          15.6    82   
393  44.0          4          97.0         52    2130          24.6    82   
394  32.0          4         135.0         84    2295          11.6    82   
395  28.0          4         120.0         79    2625          18.6    82   
396  31.0          4         119.0         82    2720          19.4    82   

     origin             name  
392       1  ford mustang gl  
393       2        vw pickup  
394       1    dodge rampage  
395       1      ford ranger  
396       1       chevy s-10

Tabular data generally, involve variables and cases. 

In ‘pandas’ data frames, each of the variables is given a name. You can refer to the variable by name in a couple of different ways:

+ We can see the variable names in a data frame by using the columns attribute of the data frame object:

In [185]:
data.columns  # This is not a function; it is an attribute of the data frame.

Index([u'year', u'time', u'sex'], dtype='object')

+ Another way to get quick information about the variables in a data frame is with the function <code>describe()</code>:

In [186]:
data.describe()  # the output from describe() is itself a data frame.

year       time
count    62.000000  62.000000
mean   1952.145161  59.924194
std      29.472881   9.916588
min    1905.000000  47.840000
25%    1924.500000  53.642500
50%    1956.500000  56.880000
75%    1975.750000  65.200000
max    2004.000000  95.000000

### Reading from the clipboard

We can also read data directly from the clipboard.

In [187]:
# we can select a part of a DataFrame of this page to build a new one.

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
bedford = pd.read_clipboard()
bedford.head()

PyperclipException: 
    Pyperclip could not find a copy/paste mechanism for your system.
    For more information, please visit https://pyperclip.readthedocs.org 

### Reading from APIs

We can read online data from public API's such as http://www.citybik.es/

In [ ]:
import pandas as pd
import seaborn as sn
d = pd.read_json('http://api.citybik.es/bicing.json')   
d.info()

In [ ]:
d.head(1)

In [ ]:
%matplotlib inline
import matplotlib.pylab as plt
a = plt.hist(d.bikes, bins=10)

### ‘pandas’ methods and ‘numpy’ operators and other functions.

There are a lot of methods that can be applied to data:

In [188]:
data['year']

0     1905
1     1908
2     1910
3     1912
4     1918
5     1920
6     1922
7     1924
8     1934
9     1935
      ... 
52    1973
53    1974
54    1976
55    1978
56    1980
57    1986
58    1992
59    1994
60    2000
61    2004
Name: year, Length: 62, dtype: int64

In [189]:
print data.year.count()    # number of non-NaN values
print data.year.mean()     # mean value 
print data.year.argmin()   # index location at which min is obtained
print data.year.min()      # min value
print data.year.sum()   # sum of values

62
1952.1451612903227
0
1905
121033


/home/alourido/anaconda3/envs/py27/lib/python2.7/site-packages/ipykernel_launcher.py:3: FutureWarning: 
The current behaviour of 'Series.argmin' is deprecated, use 'idxmin'
instead.
The behavior of 'argmin' will be corrected to return the positional
minimum in the future. For now, use 'series.values.argmin' or
'np.argmin(np.array(values))' to get the position of the minimum
row.
  This is separate from the ipykernel package so we can avoid doing imports until


It is also possible to combine ‘numpy’ operators with ‘pandas’ variables:

In [190]:
np.min(data["year"])

1905

When you encounter a function that isn’t supported by data frames, you can use ‘numpy’ functions or the special <code>apply</code> function built-into data frames.

Using the ``apply()``method, which takes an anonymous function, we can apply any function to each value in a column.

In [191]:
c = pd.DataFrame({'VarA':['aa','bb'], 'VarB':[20,30]}, index = ['Case1','Case2'])

def f(a):
    return a**2

c.VarB = c.VarB.apply(lambda d: f(d))
c

VarA  VarB
Case1   aa   400
Case2   bb   900

In [192]:
data.year.apply(np.sqrt).head(10) 

0    43.646306
1    43.680659
2    43.703547
3    43.726422
4    43.794977
5    43.817805
6    43.840620
7    43.863424
8    43.977267
9    43.988635
Name: year, dtype: float64

Alternatively, since columns are basically just arrays, we can use built-in numpy functions directly on the columns:

In [193]:
np.sqrt(data.year).head()

0    43.646306
1    43.680659
2    43.703547
3    43.726422
4    43.794977
Name: year, dtype: float64

### Aggregating

<code>groupby</code> is the ‘pandas’ way of grouping or aggregating data frames by columns.

You can construct statements that involve more than one column within a data frame. 

For instance, here’s a calculation of the mean year, separately for (grouping by) the different sexes:

In [194]:
data.groupby('sex')['year'].mean()

sex
F    1950.677419
M    1953.612903
Name: year, dtype: float64

In [195]:
res = data.groupby(['sex','year'])['time'].mean()
res.tail(20)

sex  year
M    1936    56.40
     1944    55.90
     1947    55.80
     1948    55.40
     1955    54.80
     1957    54.60
     1961    53.60
     1964    52.90
     1967    52.60
     1968    52.20
     1970    51.90
     1972    51.22
     1975    50.59
     1976    49.44
     1981    49.36
     1985    49.24
     1986    48.74
     1988    48.42
     1994    48.21
     2000    48.01
Name: time, dtype: float64

You can iterate through the result of a ``groupby`` (that returns a tuple). The first item is the column values and the second is a filtered dataframe. 

In [196]:
data.head()

year  time sex
0  1905  65.8   M
1  1908  65.6   M
2  1910  62.8   M
3  1912  61.6   M
4  1918  61.4   M

In [197]:
dic = {}
for i,j in data.groupby('sex'):
    dic[i] = j.mean()
    
print dic,

{'M': year    1953.612903
time      54.656129
dtype: float64, 'F': year    1950.677419
time      65.192258
dtype: float64}


You can group by more than one column as well: the first tuple item will itself be a tuple with the value of each column.

In [198]:
auto.head()

mpg  cylinders  displacement horsepower  weight  acceleration  year  \
0  18.0          8         307.0        130    3504          12.0    70   
1  15.0          8         350.0        165    3693          11.5    70   
2  18.0          8         318.0        150    3436          11.0    70   
3  16.0          8         304.0        150    3433          12.0    70   
4  17.0          8         302.0        140    3449          10.5    70   

   origin                       name  
0       1  chevrolet chevelle malibu  
1       1          buick skylark 320  
2       1         plymouth satellite  
3       1              amc rebel sst  
4       1                ford torino

In [203]:
auto.groupby('cylinders')

In [204]:
dic = {}
for (i,j), k in auto.groupby(['horsepower','cylinders']):
    dic[i,j] = k.mean()
    
dic

{('100', 3): mpg               23.7
 cylinders          3.0
 displacement      70.0
 horsepower       100.0
 weight          2420.0
 acceleration      12.5
 year              80.0
 origin             3.0
 dtype: float64, ('100', 4): mpg               32.9
 cylinders          4.0
 displacement     119.0
 horsepower       100.0
 weight          2615.0
 acceleration      14.8
 year              81.0
 origin             3.0
 dtype: float64, ('100', 6): mpg             1.843333e+01
 cylinders       6.000000e+00
 displacement    2.361333e+02
 horsepower      6.673340e+42
 weight          3.228067e+03
 acceleration    1.613333e+01
 year            7.380000e+01
 origin          1.000000e+00
 dtype: float64, ('102', 4): mpg               20.0
 cylinders          4.0
 displacement     130.0
 horsepower       102.0
 weight          3150.0
 acceleration      15.7
 year              76.0
 origin             2.0
 dtype: float64, ('103', 5): mpg               20.3
 cylinders          5.0
 displacemen

### Adding variables

Adding a new variable to a data frame can be done similarly to accessing a variable. For instance, here is how to create a new variable in <code>data</code> that holds the <code>time</code> converted from seconds to units of minutes:

In [205]:
data['minutes'] = data.time/60. # or data['time']/60.

By default, columns get inserted at the end. 

The <code>insert</code> function is available to insert at a particular location in the columns.

In [206]:
data.insert(1, 'mins', data.time/60.)

ValueError: cannot insert mins, already exists

In [ ]:
data['mins'] = data['time'] / 60.0

You could also, if you want, redefine an existing variable, for instance:

In [207]:
data['time'] = data.time/60.
data.head()

year      mins      time sex   minutes
0  1905  1.096667  1.096667   M  1.096667
1  1908  1.093333  1.093333   M  1.093333
2  1910  1.046667  1.046667   M  1.046667
3  1912  1.026667  1.026667   M  1.026667
4  1918  1.023333  1.023333   M  1.023333

The <code>ix[]</code> can also be used to get a subset of the dataframe:

In [208]:
kids = pd.read_csv("http://www.mosaic-web.org/go/datasets/kidsfeet.csv")
kids.shape

(39, 8)

In [209]:
kids.index

RangeIndex(start=0, stop=39, step=1)

In [210]:
# random sample of 5 cases from this data frame
rows = np.random.choice(kids.index, 5, replace=False) 

kids.ix[rows]

/home/alourido/anaconda3/envs/py27/lib/python2.7/site-packages/ipykernel_launcher.py:4: DeprecationWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#ix-indexer-is-deprecated
  after removing the cwd from sys.path.


name  birthmonth  birthyear  length  width sex biggerfoot domhand
6   Edward           2         88    26.1    9.6   B          L       R
27   David          12         87    25.5    9.5   B          R       R
38  Alisha           9         88    24.6    8.8   G          L       R
34   Peter           4         88    24.7    8.6   B          R       L
1     Lars          10         87    25.4    8.8   B          L       L

The results returned by the above methods will never contain the same case more than once (because we told the function not to sample with replacement). 

In contrast, ‘re-sampling with replacement’ replaces each case after it is dealt so that it can appear more than once in the result. You wouldn’t want to do this to select from a sampling frame, but it turns out that there are valuable statistical uses for this sort of sampling with replacement.

### Droping rows and columns.

We can delete entire rows and columns:

In [ ]:
c = pd.DataFrame({'VarA':['aa','bb'], 'VarB':[20,30]}, index = ['Case1','Case2'])
c = c.drop(['Case1'])
c

In [ ]:
c = pd.DataFrame({'VarA':['aa','bb'], 'VarB':[20,30]}, index = ['Case1','Case2'])
c = c.drop(['VarA'], axis=1)
c

### Handling missing values.

Pandas considers that values ``NaN`` and ``None`` represents missing data. 

The ``pandas.isnull`` function can be used to tell whether or not a value is missing.

In [ ]:
c = pd.DataFrame({'VarA':['aa', np.nan, 'cc'], 'VarB':[20,30,np.nan], \
                  'VarC':[1234, 3456, 6789]}, index = ['Case1','Case2','Case3'])
c

In [ ]:
empty = c.apply(lambda c: pd.isnull(c))
empty     # Empty returns a boolean dataframe

One option is to drop all rows with missing event values using the ``dropna`` function:

In [ ]:
c.dropna(subset=["VarA","VarB", "VarC"])   # Returns a dataframe copy

Another option od to use the ``fillna`` function to fillthem with empty strings:

In [ ]:
c = pd.DataFrame({'VarA':['aa', np.nan, 'cc'], 'VarB':[20,30,55], 'VarC':[1234, 3456, 6789]}, \
                 index = ['Case1','Case2','Case3'])
c = c.fillna("")
c

### Filtering

One of the main tasks when analyzing a dataset is to select rows by using simple operators. 

In [ ]:
vc = c[c.VarC > c.VarC.mean()]
vc

By using Boolean operators we can build combinde filters.

In [ ]:
vc = c[(c.VarC < c.VarC.mean()) & (c.VarC >= 1000)]
vc

A filter is a ``Series`` object with Boolean values:

In [ ]:
filt = c.VarC < c.VarC.mean()
filt

In [ ]:
vc = c[filt]
vc

We can use the ``any`` funtion to see if there is any ``True`` value in a filter:

In [ ]:
filt.any()

### Writting data

In [ ]:
bedford.to_csv("files/bedford2.csv")

In [ ]:
bedford.to_csv("files/bedford3.csv", sep="\t")

# Visualization

In [ ]:
import pandas as pd  
import matplotlib.pylab as plt
%matplotlib inline

# Load car dataset
auto = pd.read_csv("data/Auto.csv")

# create a scatterplot of weight vs "miles per galone"
auto.plot(kind='scatter', x='weight', y='mpg', color='Red', title='weight vs "miles per galone"', s=45, alpha=0.3);

# create a histogram of "miles per galone"
plt.figure()
auto.hist('mpg', alpha=0.7)
plt.title("Histogram of mpg (miles per galone)")


### EXERCISE:

Create a scatter plot using two different columns from the auto dataset

In [ ]:
import numpy as np
import pandas as pd
var = pd.DataFrame({'n': np.random.normal(size=100),
                    'g': np.random.gamma(1, size=100),
                    'p': np.random.poisson(size=100)})
var.cumsum(0).plot()

In [ ]:
var.cumsum(0).plot(secondary_y='n')

In [ ]:
fig, axes = plt.subplots(nrows=1, ncols=3, figsize=(12,4))
for i,var2 in enumerate(['n','g','p']):
    var[var2].cumsum(0).plot(ax=axes[i], title=var2)
axes[0].set_ylabel('cum sum')

In [ ]:
titanic = pd.read_excel('files/titanic.xls','titanic')

In [ ]:
titanic.groupby('pclass').survived.sum().plot(kind='bar',alpha=0.4)

In [ ]:
titanic.groupby(['sex','pclass']).survived.sum().plot(kind='barh', alpha=0.4)

In [ ]:
death_counts = pd.crosstab([titanic.pclass, titanic.sex], titanic.survived.astype(bool))
death_counts.plot(kind='bar',stacked=True, color=['lightgreen','lightblue'], grid=False)

In [ ]:
death_counts.div(death_counts.sum(1).astype(float), axis=0).plot(kind='barh', \
            stacked=True, color=['grey','gold'])

In [ ]:
titanic.fare.hist(bins=10, grid=False, alpha=0.5), titanic.fare.hist(bins=30, alpha=0.5)

There are algorithms for determining an "optimal" number of bins, each of which varies somehow with the number of observations in the data series. It is always advisable to "explore" different numbers of bins.

In [ ]:
import numpy as np
from scipy import stats

np.random.seed(0)
x = np.concatenate([stats.cauchy(-5, 1.8).rvs(500),
                    stats.cauchy(-4, 0.8).rvs(2000),
                    stats.cauchy(-1, 0.3).rvs(500),
                    stats.cauchy(2, 0.8).rvs(1000),
                    stats.cauchy(4, 1.5).rvs(500)])

# truncate values to a reasonable range
x = x[(x > -15) & (x < 15)]
# Histogram the result
import pylab as pl
H = pl.hist(x, normed=True, alpha=0.5)

In [ ]:
# Histogram with more bins
H = pl.hist(x, bins=100, alpha=0.5, normed=True)
#pl.savefig('bayesblocks2.png')

In [ ]:
def bayesian_blocks(t):
    """Bayesian Blocks Implementation

    By Jake Vanderplas.  License: BSD
    Based on algorithm outlined in http://adsabs.harvard.edu/abs/2012arXiv1207.5578S

    Parameters
    ----------
    t : ndarray, length N
        data to be histogrammed

    Returns
    -------
    bins : ndarray
        array containing the (N+1) bin edges

    Notes
    -----
    This is an incomplete implementation: it may fail for some
    datasets.  Alternate fitness functions and prior forms can
    be found in the paper listed above.
    """
    # copy and sort the array
    t = np.sort(t)
    N = t.size

    # create length-(N + 1) array of cell edges
    edges = np.concatenate([t[:1],
                            0.5 * (t[1:] + t[:-1]),
                            t[-1:]])
    block_length = t[-1] - edges

    # arrays needed for the iteration
    nn_vec = np.ones(N)
    best = np.zeros(N, dtype=float)
    last = np.zeros(N, dtype=int)

    #-----------------------------------------------------------------
    # Start with first data cell; add one cell at each iteration
    #-----------------------------------------------------------------
    for K in range(N):
        # Compute the width and count of the final bin for all possible
        # locations of the K^th changepoint
        width = block_length[:K + 1] - block_length[K + 1]
        count_vec = np.cumsum(nn_vec[:K + 1][::-1])[::-1]

        # evaluate fitness function for these possibilities
        fit_vec = count_vec * (np.log(count_vec) - np.log(width))
        fit_vec -= 4  # 4 comes from the prior on the number of changepoints
        fit_vec[1:] += best[:K]

        # find the max of the fitness: this is the K^th changepoint
        i_max = np.argmax(fit_vec)
        last[K] = i_max
        best[K] = fit_vec[i_max]
    
    #-----------------------------------------------------------------
    # Recover changepoints by iteratively peeling off the last block
    #-----------------------------------------------------------------
    change_points =  np.zeros(N, dtype=int)
    i_cp = N
    ind = N
    while True:
        i_cp -= 1
        change_points[i_cp] = ind
        if ind == 0:
            break
        ind = last[ind - 1]
    change_points = change_points[i_cp:]

    return edges[change_points]

In [ ]:
# plot a standard histogram in the background, with alpha transparency
H1 = plt.hist(x, bins=200, histtype='stepfilled', alpha=0.3, normed=True)
# plot an adaptive-width histogram on top
H2 = plt.hist(x, bins=bayesian_blocks(x), color='red', histtype='step', normed=True)
#pl.savefig('bayesblocks3.png')

A different way of visualizing the distribution of data is the ``boxplot``, which is a display of common quantiles; these are typically the quartiles and the lower and upper 5 percent values.

In [ ]:
titanic.boxplot(column='fare', by='pclass', grid=False)

In [ ]:
bp = titanic.boxplot(column='age', by='pclass', grid=False)
for i in [1,2,3]:
    y = titanic.age[titanic.pclass==i].dropna()
    # Add some random "jitter" to the x-axis
    x = np.random.normal(i, 0.04, size=len(y))
    plt.plot(x, y, 'r.', alpha=0.2)

Scatterplots

In [ ]:
baseball = pd.read_csv("files/baseball.csv")
baseball.head()

In [ ]:
plt.scatter(baseball.ab, baseball.h)
plt.xlim(0, 700); plt.ylim(0, 200)

In [ ]:
plt.scatter(baseball.ab, baseball.h, s=baseball.hr*10, alpha=0.3)
plt.xlim(0, 700); plt.ylim(0, 200)

In [ ]:
plt.scatter(baseball.ab, baseball.h, c=baseball.hr, s=baseball.hr*10, alpha=0.5, cmap='hot')
plt.xlim(0, 700); plt.ylim(0, 200);

To view scatterplots of a large numbers of variables simultaneously, we can use the scatter_matrix function that was recently added to Pandas. It generates a matrix of pair-wise scatterplots, optiorally with histograms or kernel density estimates on the diagonal.

In [ ]:
_ = pd.plotting.scatter_matrix(baseball.loc[:,'r':'sb'], figsize=(12,8), diagonal='kde')

## Python Patterns and Anti-patterns

#### Test for “truthy” and “falsy” values:

+ Checking for truth doesn't tie the conditional expression to the type of object being checked.
+ Checking for truth clearly shows the code's intention rather than drawing attention to a specific outcome.

In [ ]:
# GOOD
name = 'Safe'
pets = ['Dog', 'Cat', 'Hamster']
owners = {'Safe': 'Cat', 'George': 'Dog'}
if name and pets and owners:
    print('We have pets!')

# NOT SO GOOD
if name != '' and len(pets) > 0 and owners != {}:
    print('We have pets!')

####  Use <code>in</code> where possible:

+ Using ``in`` to check if an item is in a sequence is clear and concise.
+ Can be used on lists, dicts (keys), sets, strings, and your own classes by implementing the <code>__contains__<code> special method.


In [ ]:
# GOOD
name = 'Safe Hammad'
if 'H' in name:
    print('This name has an H in it!')

# NOT SO GOOD
name = 'Safe Hammad'
if name.find('H') != -1:
    print('This name has an H in it!')


+ Using ``in`` to for iteration over a sequence is clear and concise.
+ Can be used on lists, dicts (keys), sets, strings, and your own classes by implementing the <code>__iter__</code> special method.

In [ ]:
# GOOD
pets = ['Dog', 'Cat', 'Hamster']
for pet in pets:
    print('A', pet, 'can be very cute!')
    
# NOT SO GOOD
pets = ['Dog', 'Cat', 'Hamster']
i = 0
while i < len(pets):
    print('A', pets[i], 'can be very cute!')
    i += 1


####  Swap values without temp variable: Avoids polluting namespace with temp variable used only once.

In [ ]:
# GOOD
a, b = 5, 6
print(a, b) # 5, 6
a, b = b, a
print(a, b) # 6, 5
 
# NOT SO GOOD
a, b = 5, 6
print(a, b) # 5, 6
 
temp = a
a = b
b = temp
print(a, b) # 6, 5


####  Enumerate! 

In [ ]:
# GOOD
names = ['Safe', 'George', 'Mildred']
for i, name in enumerate(names):
    print(i, name) # 0 Safe, 1 George etc.

# NOT SO GOOD
names = ['Safe', 'George', 'Mildred']
count = 0
for name in names:
    print(i, name) # 0 Safe, 1 George etc.
    count += 1

Build lists (dictionaries) using list (dictionaries) comprehensions!

In [ ]:
# GOOD
data = [7, 20, 3, 15, 11]
result = [i * 3 for i in data if i > 10]
print(result) # [60, 45, 33]

# NOT SO GOOD (MOST OF THE TIME)
data = [7, 20, 3, 15, 11]
result = [] 
for i in data:
    if i > 10:
        result.append(i * 3)
print(result) # [60, 45, 33]

####  Create dict from keys and values using zip.

In [ ]:
# GOOD
keys = ['Safe', 'Bob', 'Thomas']
values = ['Hammad', 'Builder', 'Engine'] 
d = dict(zip(keys, values))
print(d) # {'Bob': 'Builder','Safe': 'Hammad','Thomas': 'Engine'}

# NOT SO GOOD
keys = ['Safe', 'Bob', 'Thomas']
values = ['Hammad', 'Builder', 'Engine']
d = {}
for i, key in enumerate(keys):
    d[key] = values[i]
print(d) # {'Bob': 'Builder','Safe': 'Hammad','Thomas': 'Engine'}


In [ ]:
keys = ['a', 'b', 'c']
values = [1, 2, 3]
dictionary = dict(zip(keys, values))
print dictionary

If you have to deal with millions of (immutable) strings, use a list to hold the substrings and join them with a string method:

In [ ]:
def genChar1():
    from random import randint
    s = ''
    for i in xrange(10000000):
        s += 'a'
    return s

%timeit genChar1()

In [ ]:
def genChar2():
    from random import randint
    s = []
    for i in xrange(100000000):
        s.append('a')
    return ''.join(s)

%timeit genChar2()

If you are interested in some values computed from a whole list, you can use generator comprehension to avoid the list:

In [ ]:
%timeit sum([x * x for x in xrange(100000000)])

In [ ]:
%timeit sum(x * x for x in xrange(100000000))

If you need to search in items, dictionaries and sets are preferable to lists.

In [ ]:
a = range(1000000)
b = set(range(1000000))
%timeit 9 in a
%timeit 9 in b

## Python paralell processing

In [ ]:
def sample(n):
    from numpy.random import rand
    return(rand(n) **2.0 + rand(n)**2.0 <=1).sum()

n = 1000000

from IPython.parallel import Client
rc = Client()
v = rc[:]
print len(v), 'cores'
with v.sync_imports():
    from numpy.random import random

4.0 * sum(v.map_sync(sample, [n] * len(v)))/(n*len(v))

# Python online resources

+ Python Practice Book: [Link](http://anandology.com/python-practice-book/index.html)
+ Python Online Course at Code Academy: [Link](http://www.codecademy.com/en/tracks/python)
+ Think Python: How to Think Like a Computer Scientist Book: [Link](http://greenteapress.com/thinkpython/index.html)